From adaf9d2c863cf5d18bfd8d3918266d25aa07786c Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:37:26 -0400
Subject: [virt] kvm/nvmx: reset rflags register cache during nested vmentry

Message-id: <1406040016-3289-52-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86130
O-Subject: [RHEL7 PATCH v2 051/221] nVMX: reset rflags register cache during nested vmentry.
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Gleb Natapov <gleb@redhat.com>

During nested vmentry into vm86 mode a vcpu state is found to be incorrect
because rflags does not have VM flag set since it is read from the cache
and has L1's value instead of L2's. If emulate_invalid_guest_state=1 L0
KVM tries to emulate it, but emulation does not work for nVMX and it
never should happen anyway. Fix that by using vmx_set_rflags() to set
rflags during nested vmentry which takes care of updating register cache.

Signed-off-by: Gleb Natapov <gleb@redhat.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit 63fbf59f8a31548e64cdc3adc5d0997be0486b27)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index abe861e..17caa24 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -7419,7 +7419,7 @@ static void prepare_vmcs02(struct kvm_vcpu *vcpu, struct vmcs12 *vmcs12)
   vmcs12->guest_interruptibility_info);
  vmcs_write32(GUEST_SYSENTER_CS, vmcs12->guest_sysenter_cs);
  kvm_set_dr(vcpu, 7, vmcs12->guest_dr7);
- vmcs_writel(GUEST_RFLAGS, vmcs12->guest_rflags);
+ vmx_set_rflags(vcpu, vmcs12->guest_rflags);
  vmcs_writel(GUEST_PENDING_DBG_EXCEPTIONS,
   vmcs12->guest_pending_dbg_exceptions);
  vmcs_writel(GUEST_SYSENTER_ESP, vmcs12->guest_sysenter_esp);
-- 
1.7.1